### ISSUE TITLE
Failed tests (10719): TestMetricsRecording leaks (*Server).{sendResponses, readRequests}

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#3664]
The following test appears to have failed:

[#10719](https://circleci.com/gh/cockroachdb/cockroach/10719):

```
I0107 09:19:38.033725 743 stopper.go:264  draining; tasks left:
1      ts/db.go:98
I0107 09:19:38.132827 743 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestMetricsRecording (1.13s)
PASS
Too many goroutines running after tests.
1 instances of:
github.com/cockroachdb/cockroach/rpc.(*Server).sendResponses(0xc820ac2d80, 0x7fe06ec289c0, 0xc82021a300, 0xc820b7ac60)
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:300 +0xb5
github.com/cockroachdb/cockroach/rpc.(*Server).ServeHTTP.func1(0xc820ac2d80, 0x7fe06ec289c0, 0xc82021a300, 0xc820b7ac60)
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:207 +0x4d
created by github.com/cockroachdb/cockroach/rpc.(*Server).ServeHTTP
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:208 +0xa6c
1 instances of:
github.com/cockroachdb/cockroach/storage.(*Replica).addWriteCmd(0xc82052fd40, 0x7fe06dbbeff8, 0xc820406840, 0x14271a55cc4f30a6, 0x1, 0x100000001, 0x1, 0x1, 0x0, 0xc8202cc480, ...)
	/go/src/github.com/cockroachdb/cockroach/storage/replica.go:926 +0x7be
--
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:238 +0x27e
github.com/cockroachdb/cockroach/rpc.(*Server).ServeHTTP.func2(0xc820ac2d80, 0x7fe06dbbf8a8, 0xc820590000, 0x7fe06ec289c0, 0xc82021a300, 0xc8202dc230, 0xc820b7ac60)
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:210 +0x6f
created by github.com/cockroachdb/cockroach/rpc.(*Server).ServeHTTP
	/go/src/github.com/cockroachdb/cockroach/rpc/server.go:213 +0xadc
FAIL	github.com/cockroachdb/cockroach/server	27.436s
=== RUN   TestNodeStatusMonitor
--- PASS: TestNodeStatusMonitor (0.01s)
=== RUN   TestNodeStatusRecorder
--- PASS: TestNodeStatusRecorder (0.02s)
=== RUN   TestRuntimeStatRecorder
I0107 09:19:22.761298 769 server/status/runtime.go:137  runtime stats: 11 goroutines, 4.92MiB active, 50000000.00cgo/sec, 781580.00/156310.00 %(u/s)time, 8580.91 %gc (2x)
--- PASS: TestRuntimeStatRecorder (0.00s)
=== RUN   TestNodeEventFeed
--- PASS: TestNodeEventFeed (0.00s)
=== RUN   TestServerNodeEventFeed
--
I0107 09:21:02.702000 931 storage/replica_command.go:1198  range 1: new leader lease replica {1 1 1} 1970-01-01 00:00:00 +0000 UTC +1.000s
I0107 09:21:02.702623 931 storage/replica.go:1619  gossiping system config from store 1, range 1
I0107 09:21:02.703157 931 storage/replica.go:1559  gossiping cluster id  from store 1, range 1
I0107 09:21:02.703573 931 storage/replica.go:1568  gossiping sentinel from store 1, range 1
I0107 09:21:02.704696 931 storage/replica.go:1574  gossiping first range from store 1, range 1
panic: test timed out after 5m0s

goroutine 3219 [running]:
testing.startAlarm.func1()
	/tmp/workdir/go/src/testing/testing.go:703 +0x16b
created by time.goFunc
	/tmp/workdir/go/src/time/sleep.go:129 +0x6e

goroutine 1 [chan receive, 4 minutes]:
testing.RunTests(0x1a6a690, 0x2117360, 0xdb, 0xdb, 0x1)
	/tmp/workdir/go/src/testing/testing.go:562 +0xafa
testing.(*M).Run(0xc820034f10, 0x1)
	/tmp/workdir/go/src/testing/testing.go:494 +0xe5
github.com/cockroachdb/cockroach/util/leaktest.TestMainWithLeakCheck(0xc820034f10)
	/go/src/github.com/cockroachdb/cockroach/util/leaktest/leaktest.go:34 +0x2f
github.com/cockroachdb/cockroach/storage_test.TestMain(0xc820034f10)
	/go/src/github.com/cockroachdb/cockroach/storage/main_test.go:36 +0x34
main.main()
	github.com/cockroachdb/cockroach/storage/_test/_testmain.go:498 +0x20a

goroutine 17 [syscall, 4 minutes, locked to thread]:
runtime.goexit()
	/tmp/workdir/go/src/runtime/asm_amd64.s:1721 +0x1

goroutine 20 [chan receive]:
github.com/cockroachdb/cockroach/util/log.(*loggingT).flushDaemon(0x21725a0)
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:1029 +0x76
created by github.com/cockroachdb/cockroach/util/log.init.1
	/go/src/github.com/cockroachdb/cockroach/util/log/clog.go:610 +0x117

goroutine 3180 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d40e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3195 [select, 2 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:621 +0x447
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8202f8a00)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3183 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d4140)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3184 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d4160)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3185 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d41a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3182 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d4120)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3181 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*baseQueue).processLoop.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/queue.go:319 +0x427
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8201d4100)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3192 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/util.ListenAndServe.func2()
	/go/src/github.com/cockroachdb/cockroach/util/net.go:88 +0x7b
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8202f89a0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3194 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).Start.func3()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:556 +0x15d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8202f89e0)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3190 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open.func1(0xc82018e1e0)
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:123 +0x5b
created by github.com/cockroachdb/cockroach/storage/engine.(*RocksDB).Open
	/go/src/github.com/cockroachdb/cockroach/storage/engine/rocksdb.go:124 +0x581

goroutine 3196 [select, 2 minutes]:
github.com/cockroachdb/cockroach/storage.(*Store).startGossip.func2()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:640 +0x447
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8202f8a20)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3189 [select, 4 minutes]:
github.com/cockroachdb/cockroach/storage.(*StorePool).start.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store_pool.go:213 +0x208
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc8202f8340)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3188 [chan receive, 4 minutes]:
github.com/cockroachdb/cockroach/storage.TestMaybeRemove(0xc82008e000)
	/go/src/github.com/cockroachdb/cockroach/storage/store_test.go:1648 +0x78c
testing.tRunner(0xc82008e000, 0x21181a0)
	/tmp/workdir/go/src/testing/testing.go:456 +0xdd
created by testing.RunTests
	/tmp/workdir/go/src/testing/testing.go:561 +0xaa4

goroutine 3193 [select]:
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/go/src/github.com/cockroachdb/cockroach/storage/store.go:1614 +0xb0c
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc820301060)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70

goroutine 3191 [IO wait, 4 minutes]:
net.runtime_pollWait(0x7f1aa46df470, 0x72, 0x7f1aa46d2028)
	/tmp/workdir/go/src/runtime/netpoll.go:157 +0x63
net.(*pollDesc).Wait(0xc8200ccbc0, 0x72, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:73 +0x56
net.(*pollDesc).WaitRead(0xc8200ccbc0, 0x0, 0x0)
	/tmp/workdir/go/src/net/fd_poll_runtime.go:78 +0x44
net.(*netFD).accept(0xc8200ccb60, 0x0, 0x7f1aa57385d8, 0xc8202f8b40)
	/tmp/workdir/go/src/net/fd_unix.go:408 +0x2f6
net.(*TCPListener).AcceptTCP(0xc820082048, 0x60ce50, 0x0, 0x0)
	/tmp/workdir/go/src/net/tcpsock_posix.go:254 +0x77
--
	/go/src/github.com/cockroachdb/cockroach/util/net.go:76 +0x21d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8200cc000, 0xc820303d40)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:89 +0x60
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:90 +0x70
FAIL	github.com/cockroachdb/cockroach/storage	300.064s
=== RUN   TestBatchBasics
I0107 09:21:08.914573 993 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.00s)
=== RUN   TestBatchGet
I0107 09:21:08.918510 993 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.00s)
=== RUN   TestBatchMerge
I0107 09:21:08.923576 993 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
those look like maybe they simply ought to be exempt by leakcheck? The second is started by

```go
	go func() {
		s.readRequests(conn, codec, authHook, responses)
		codec.Close()
		conn.Close()
	}()

```

so it doesn't seem to have stopper protection. @bdarnell?

**Comment by bdarnell** [comment]
They definitely shouldn't be exempt from `leakcheck`. The client uses the stopper and shuts down the client side of the connection; this should be enough to allow the server-side goroutines to detect the failure and shut down within the `leakcheck` grace period. We may need to make that grace period longer but if these goroutines aren't going away I think it's something we can/should fix. 

Alternately, we could use `RunWorker` for these goroutines (similar to the way that @petermattis is adding `RunTask` to other rpc server code).

**Comment by petermattis** [comment]
If you look at the full stack traces, the `readRequests` goroutine is blocked in the defer waiting for the `wg.Wait()` to return. Which must mean that a request was received and is currently blocked in a `meth.handler`, but we're not showing any goroutine handling a request. What gives?

**Comment by petermattis** [comment]
Never mind, I can't read. We have a goroutine blocked in `Replica.addWriteCmd`. I'm not familiar with this code, but I imagine range isn't applying the command because the stopper has been stopped.

**Comment by tschottdorf** [comment]
@petermattis is there reason to assume that your recent stopper PR has squashed this as well?

**Comment by petermattis** [comment]
It is possible #3683 fixed this. I can imagine a scenario where the server received a request and proposed it to raft but that request was never executed because raft was shut down (the stopper is stopped in the backtraces). I have failed at reproducing this error after 13000 runs using `make stressrace STRESSFLAGS="-p 24" PKG=./server TESTS=TestMetricsRecording`. I'm going to revert my stopper change and see if I can reproduce.

**Comment by petermattis** [comment]
And now 10K runs without my stopper change and no failures. Let me try with the exact change that failed.

**Comment by petermattis** [comment]
Grrr, 20K runs without failure at 30cabc0f47e5c9927efb095bb4ffd0a59589f347 which is the SHA that failed on CircleCI. Not having any success at reproducing this bug.

**Comment by petermattis** [comment]
Ah-ha! `TestMetricsRecording` was a red-herring! I haven't narrowed down which is the problematic test, but I can reproduce the exact goroutine leakage at 30cabc0f47e5c9927efb095bb4ffd0a59589f347 using `make stress PKG=./server`. That same invocation does not produce any failures (after running for an hour) at 99d1bde339461991849ae56211e6398045e9136e. I'm going to close this bug and claim it was fixed by #3683.
